lib/extensions/postgres_cdc_rls/subscriptions_checker.ex

```diff
@@ -57,33 +68,133 @@ defmodule Extensions.PostgresCdcRls.SubscriptionsChecker do
+
+    new_delete_queue =
+      if length(ids) > 0 do
+        q =
+          Enum.reduce(ids, delete_queue.queue, fn id, acc ->
+            if :queue.member(id, acc), do: acc, else: :queue.in(id, acc)
+          end)
+
+        %{
+          ref: check_delete_queue(),
+          queue: q
+        }
+      else
+        delete_queue
+      end
 
-          _ ->
-            Logger.error("Detected phantom subscriber")
-            :ets.delete(tid, pid)
-            [UUID.string_to_binary!(postgres_id) | acc]
+    {:noreply, %{state | check_active_pids: check_active_pids(), delete_queue: new_delete_queue}}
+  end
+
+  def handle_info(:check_delete_queue, %State{delete_queue: %{ref: ref, queue: q}} = state) do
+    H.cancel_timer(ref)
+
+    new_queue =
+      if !:queue.is_empty(q) do
+        {ids, q1} = H.queue_take(q, @max_delete_records)
+        Logger.error("Delete #{length(ids)} phantom subscribers from db")
+
+        case Subscriptions.delete_multi(state.conn, ids) do
+          {:ok, _} ->
+            q1
+
+          {:error, reason} ->
+            Logger.error("delete phantom subscriptions from the queue failed: #{inspect(reason)}")
+            q
         end
+      else
+        q
       end
-      |> :ets.foldl([], tid)
 
-    if length(ids) > 0 do
-      Subscriptions.delete_multi(state.conn, ids)
-    end
+    new_ref = if !:queue.is_empty(new_queue), do: check_delete_queue(), else: ref
 
-    {:noreply, %{state | check_active_pids: check_active_pids()}}
+    {:noreply, %{state | delete_queue: %{ref: new_ref, queue: new_queue}}}
   end
 
   ## Internal functions
 
+  @spec pop_not_alive_pids([pid()], :ets.tid()) :: [Ecto.UUID.t()]
+  def pop_not_alive_pids(pids, tid) do
+    Enum.reduce(pids, [], fn pid, acc ->
+      case :ets.lookup(tid, pid) do
+        [] ->
+          Logger.error("Can't find pid in subscribers table: #{inspect(pid)}")
+          acc
+
+        results ->
+          for {^pid, postgres_id, _ref, _node} <- results do
+            Logger.error(
+              "Detected phantom subscriber #{inspect(pid)} with postgres_id #{inspect(postgres_id)}"
+            )
+
+            :ets.delete(tid, pid)
+            UUID.string_to_binary!(postgres_id)
+          end ++ acc
+      end
+    end)
+  end
+
+  @spec subscribers_by_node(:ets.tid()) :: %{node() => MapSet.t(pid())}
+  def subscribers_by_node(tid) do
+    fn {pid, _postgres_id, _ref, node}, acc ->
+      set =
+        if Map.has_key?(acc, node) do
+          MapSet.put(acc[node], pid)
+        else
+          MapSet.new([pid])
+        end
+
+      Map.put(acc, node, set)
+    end
+    |> :ets.foldl(%{}, tid)
+  end
+
+  @spec not_alive_pids_dist(%{node() => MapSet.t(pid())}) :: [pid()] | []
+  def not_alive_pids_dist(pids) do
+    Enum.reduce(pids, [], fn {node, pids}, acc ->
+      if node == node() do
+        acc ++ not_alive_pids(pids)
+      else
+        case :rpc.call(node, __MODULE__, :not_alive_pids, [pids], 15_000) do
```

Comment:
>If SubscriptionManager is doing Process.monitor/1 and it is handling :DOWN messages and cleaning up the subscription, why does SubscriptionChecker need to periodically check for pids on remote nodes? Shouldn't the SubscriptionManager handle it since it is monitoring remote processes?

for some reason, we don't always receive the `down` message, this is why there exists checker

>We can consider using Registry instead of using :ets directly and having to worry about cleaning up.

how do we delete the record from the subscription table in this case?

Reply:
also, the current code in checker doesn't add the tenant id to the log, it's hard to debug

Reply:
> how do we delete the record from the subscription table in this case?

I meant we don't have to worry about cleaning up :ets records b/c Registry will handle it but we will still have to monitor process and then delete it from table on :DOWN.

Reply:
This PR is related only to updates in checker. Does it make current behavior worse than it is now? What is the reason for blocking it?

Reply:
@abc3 well it was blocked when i wrote "reviewing..." b/c it was approved and i wanted to look through it. this review was merely a comment. i'll approve it so it's not blocked.